// @zerodevapp/sdk/dist/src/zerodev/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAIN_ID_TO_TRANSAK_NAME = exports.CHAIN_ID_TO_INFURA_NAME = exports.WALLET_CONNECT_RELAY_URL = exports.WALLET_CONNECT_PROJECT_ID = exports.TRANSAK_PRODUCTION_API_KEY = exports.TRANSAK_STAGING_API_KEY = exports.INFURA_API_KEY = exports.ACCOUNT_FACTORY_ADDRESS = exports.PAYMASTER_ADDRESS = exports.ENTRYPOINT_ADDRESS = exports.BUNDLER_URL = exports.PAYMASTER_ABI = exports.PAYMASTER_URL = exports.BACKEND_URL = void 0;
exports.BACKEND_URL = (_a = process.env.REACT_APP_ZERODEV_BACKEND_URL) !== null && _a !== void 0 ? _a : 'https://backend-vikp.onrender.com';
exports.PAYMASTER_URL = (_b = process.env.REACT_APP_ZERODEV_PAYMASTER_URL) !== null && _b !== void 0 ? _b : 'https://staging-paymaster.onrender.com';
var paymaster_abi_json_1 = require("./abi/paymaster_abi.json");
Object.defineProperty(exports, "PAYMASTER_ABI", { enumerable: true, get: function () { return __importDefault(paymaster_abi_json_1).default; } });
exports.BUNDLER_URL = {
    '1': 'https://mainnet-bundler-v2.onrender.com/rpc',
    '5': 'https://goerli-bundler-v2.onrender.com/rpc',
    // '137': 'https://polygon-bundler-v2.onrender.com/rpc',
    '137': 'http://localhost:3000/rpc',
    '43114': 'https://avalanche-bundler-v2.onrender.com/rpc',
    '80001': 'https://mumbai-bundler-v2.onrender.com/rpc',
};
exports.ENTRYPOINT_ADDRESS = '0x0F46c65C17AA6b4102046935F33301f0510B163A';
exports.PAYMASTER_ADDRESS = '0x95341fe310FcDcA0d08c7b263773963ff4Bc3439';
exports.ACCOUNT_FACTORY_ADDRESS = '0x96fc755cB191B0eF516942B3Bd8aAF50E22574D6';
exports.INFURA_API_KEY = 'f36f7f706a58477884ce6fe89165666c';
exports.TRANSAK_STAGING_API_KEY = '135ef8d4-982d-4ec2-a3b2-2a2263ffbca2';
exports.TRANSAK_PRODUCTION_API_KEY = 'fb1517e3-e91c-4694-8de0-7d2a8078f3cf';
exports.WALLET_CONNECT_PROJECT_ID = '9832ea3eefe6c1b75a689ed0c90ce085';
exports.WALLET_CONNECT_RELAY_URL = 'wss://relay.walletconnect.com';
exports.CHAIN_ID_TO_INFURA_NAME = {
    '1': 'mainnet',
    '5': 'goerli',
    '137': 'polygon-mainnet',
    '80001': 'polygon-mumbai',
    '10': 'optimism-mainnet',
    '420': 'optimism-goerli',
    '42161': 'arbitrum-mainnet',
    '421613': 'arbitrum-goerli',
    '43114': 'avalanche-mainnet',
    '43113': 'avalanche-fuji',
    '1313161554': 'aurora-mainnet',
    '1313161555': 'aurora-testnet',
};
exports.CHAIN_ID_TO_TRANSAK_NAME = {
    '1': 'ethereum',
    '5': 'goerli',
    '137': 'polygon',
    '80001': 'mumbai',
    '43114': 'avaxcchain',
    '10': 'optimism',
    '42161': 'arbitrum',
};
