// @zerodevapp/sdk/dist/src/ERC4337EthersSigner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC4337EthersSigner = void 0;
const properties_1 = require("@ethersproject/properties");
const abstract_signer_1 = require("@ethersproject/abstract-signer");
const ethers_1 = require("ethers");
const types_1 = require("./types");
class ERC4337EthersSigner extends abstract_signer_1.Signer {
    // TODO: we have 'erc4337provider', remove shared dependencies or avoid two-way reference
    constructor(config, originalSigner, erc4337provider, httpRpcClient, smartAccountAPI) {
        super();
        this.config = config;
        this.originalSigner = originalSigner;
        this.erc4337provider = erc4337provider;
        this.httpRpcClient = httpRpcClient;
        this.smartAccountAPI = smartAccountAPI;
        (0, properties_1.defineReadOnly)(this, 'provider', erc4337provider);
    }
    delegateCopy() {
        // copy the account API except with delegate mode set to true
        const delegateAccountAPI = Object.assign({}, this.smartAccountAPI);
        Object.setPrototypeOf(delegateAccountAPI, Object.getPrototypeOf(this.smartAccountAPI));
        delegateAccountAPI.delegateMode = true;
        return new ERC4337EthersSigner(this.config, this.originalSigner, this.erc4337provider, this.httpRpcClient, delegateAccountAPI);
    }
    // This one is called by Contract. It signs the request and passes in to Provider to be sent.
    async sendTransaction(transaction) {
        var _a, _b, _c, _d, _e;
        // `populateTransaction` internally calls `estimateGas`.
        // Some providers revert if you try to call estimateGas without the wallet first having some ETH,
        // which is going to be the case here if we use paymasters.  Therefore we set the gas price to
        // 0 to ensure that estimateGas works even if the wallet has no ETH.
        if (transaction.maxFeePerGas || transaction.maxPriorityFeePerGas) {
            transaction.maxFeePerGas = 0;
            transaction.maxPriorityFeePerGas = 0;
        }
        else {
            transaction.gasPrice = 0;
        }
        const tx = await this.populateTransaction(transaction);
        await this.verifyAllNecessaryFields(tx);
        let userOperation;
        userOperation = await this.smartAccountAPI.createSignedUserOp({
            target: (_a = tx.to) !== null && _a !== void 0 ? _a : '',
            data: (_c = (_b = tx.data) === null || _b === void 0 ? void 0 : _b.toString()) !== null && _c !== void 0 ? _c : '',
            value: tx.value,
            gasLimit: tx.gasLimit,
            maxFeePerGas: tx.maxFeePerGas,
            maxPriorityFeePerGas: tx.maxPriorityFeePerGas,
        });
        const transactionResponse = await this.erc4337provider.constructUserOpTransactionResponse(userOperation);
        // Invoke the transaction hook
        (_e = (_d = this.config.hooks) === null || _d === void 0 ? void 0 : _d.transactionStarted) === null || _e === void 0 ? void 0 : _e.call(_d, {
            hash: transactionResponse.hash,
            from: tx.from,
            to: tx.to,
            value: tx.value || 0,
            sponsored: userOperation.paymasterAndData !== '0x',
            module: (0, types_1.getModuleInfo)(tx),
        });
        try {
            await this.httpRpcClient.sendUserOpToBundler(userOperation);
        }
        catch (error) {
            // console.error('sendUserOpToBundler failed', error)
            throw this.unwrapError(error);
        }
        // TODO: handle errors - transaction that is "rejected" by bundler is _not likely_ to ever resolve its "wait()"
        return transactionResponse;
    }
    unwrapError(errorIn) {
        var _a;
        if (errorIn.body != null) {
            const errorBody = JSON.parse(errorIn.body);
            let paymasterInfo = '';
            let failedOpMessage = (_a = errorBody === null || errorBody === void 0 ? void 0 : errorBody.error) === null || _a === void 0 ? void 0 : _a.message;
            if ((failedOpMessage === null || failedOpMessage === void 0 ? void 0 : failedOpMessage.includes('FailedOp')) === true) {
                // TODO: better error extraction methods will be needed
                const matched = failedOpMessage.match(/FailedOp\((.*)\)/);
                if (matched != null) {
                    const split = matched[1].split(',');
                    paymasterInfo = `(paymaster address: ${split[1]})`;
                    failedOpMessage = split[2];
                }
            }
            const error = new Error(`The bundler has failed to include UserOperation in a batch: ${failedOpMessage} ${paymasterInfo})`);
            error.stack = errorIn.stack;
            return error;
        }
        return errorIn;
    }
    async estimateGas(transaction) {
        var _a, _b, _c;
        const tx = await (0, properties_1.resolveProperties)(this.checkTransaction(transaction));
        let userOperation;
        userOperation = await this.smartAccountAPI.createUnsignedUserOp({
            target: (_a = tx.to) !== null && _a !== void 0 ? _a : '',
            data: (_c = (_b = tx.data) === null || _b === void 0 ? void 0 : _b.toString()) !== null && _c !== void 0 ? _c : '',
            value: tx.value,
            maxFeePerGas: tx.maxFeePerGas,
            maxPriorityFeePerGas: tx.maxPriorityFeePerGas,
        });
        const gasInfo = await this.httpRpcClient.estimateUserOpGas(userOperation);
        return ethers_1.BigNumber.from(gasInfo.preVerificationGas).add(ethers_1.BigNumber.from(gasInfo.verificationGas)).add(ethers_1.BigNumber.from(gasInfo.callGasLimit));
    }
    async verifyAllNecessaryFields(transactionRequest) {
        if (transactionRequest.to == null) {
            throw new Error('Missing call target');
        }
        if (transactionRequest.data == null && transactionRequest.value == null) {
            // TBD: banning no-op UserOps seems to make sense on provider level
            throw new Error('Missing call data or value');
        }
    }
    connect(provider) {
        throw new Error('changing providers is not supported');
    }
    async getAddress() {
        if (this.address == null) {
            this.address = await this.erc4337provider.getSenderAccountAddress();
        }
        return this.address;
    }
    async signMessage(message) {
        return await this.originalSigner.signMessage(message);
    }
    async signTransaction(transaction) {
        throw new Error('not implemented');
    }
    async signUserOperation(userOperation) {
        const message = await this.smartAccountAPI.getUserOpHash(userOperation);
        return await this.originalSigner.signMessage(message);
    }
}
exports.ERC4337EthersSigner = ERC4337EthersSigner;
